### PR TITLE
[19] [20] [Integrate] [Backend] As a logged-in user, I can search repositories to notify

### LINKED ISSUE #19: [Backend] As a logged-in user, I can search repositories to notify
## Why
To support the logged-in user to select repositories they want for notifications, we need to create a repositories preferences screen

## Acceptance Criteria

- We need to get the list of subscribed (watching) repositories to show on the list:

```
Header:
Authorization: token {access_token}
Accept: application/vnd.github.v3+json

Parameter: per_page: 100

Method: Get

Path: https://api.github.com/users/subscriptions
```

Design: https://www.figma.com/file/jCSHtb5pycYvT57B5MZaSu/Github-Notifications?node-id=7%3A108

### LINKED ISSUE #20: [Integrate] As a logged-in user, I can search repositories to notify
## Why

To support the logged-in user to select repositories they want for notifications, we need to create a repositories preferences screen

## Acceptance Criteria

- After we get a list of watching repositories from API. 
- Show the list on the UI
- The user can search by the name of the repository.
- We filter the list every time the user input a character.
- When the user press on the done button, we move back to the previous screen.

Design: https://www.figma.com/file/jCSHtb5pycYvT57B5MZaSu/Github-Notifications?node-id=7%3A108

### PR DESCRIPTION
**Pull request by vnntsu**
resolve #19 
resolve #20

## What happened 👀

- After we get a list of watching repositories from API.
- Show the list on the UI
- The user can search by the name of the repository.
- We filter the list every time the user input a character.
- When the user press on the done button, we move back to the previous screen.
- We need to get the list of subscribed (watching) repositories to show on the list:

```
Header:
Authorization: token {access_token}
Accept: application/vnd.github.v3+json

Parameter: per_page: 100

Method: Get

Path: https://api.github.com/users/user/subscriptions
```
 
## Insight 📝

- Create `RepositoryService` to get subscribing repo

### CODE REVIEW

File: Source/Services/Network/Services/RepositoryService.swift [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        self.networkAPI = networkAPI
+        self.userSession = userSession
+    }
+
+    func getRepositories() -> AnyPublisher<[Repository], Error> {
+//        networkAPI.performRequest(
+//            with: RepositoryRequestConfiguration.getRepositories(userSession.username ?? "")
+//        )
+
+        Future { promise in
```
Comment (blyscuit) [inline Source/Services/Network/Services/RepositoryService.swift:35]: As task #19 is backend task, is there a bug that makes you need to use a mock data?

File: Source/Module/SearchRepository/SearchRepoScreen.swift [hunk 2]
```diff
@@ -10,31 +10,23 @@ import SwiftUI
...
                 .paddingLeft8()
 
             ScrollView(.vertical) {
                 LazyVStack(alignment: .leading) {
-                    ForEach(Array(RepositoryCellViewModel.dummy.enumerated()), id: \.offset) { index, item in
+                    ForEach(viewModel.filteredItems) { item in
                         RepositoryCell(viewModel: item, isNotifying: notifiedReposObservable.getIsNotifying(item))
-                            .background(
-                                index % 2 == 0 ? Color.white : Color.gray.opacity(0.1)
-                            )
```
Comment (blyscuit) [inline Source/Module/SearchRepository/SearchRepoScreen.swift:37]: Is it possible to keep these line with
```
ForEach(viewModel.filteredItems.enumerated()) { index, item in
```

Comment (blyscuit) [reply Source/Module/SearchRepository/SearchRepoScreen.swift:37]: I see, there's a bug with catalyst when search bar is active and scrolling.

Comment (blyscuit) [reply Source/Module/SearchRepository/SearchRepoScreen.swift:37]: This block seems to work fine
```

            ScrollView(.vertical) {
                LazyVStack(alignment: .leading) {
                    ForEach(Array(zip(viewModel.filteredItems.indices, viewModel.filteredItems)), id: \.0) { index, item in
                        RepositoryCell(viewModel: item, isNotifying: notifiedReposObservable.getIsNotifying(item))
                            .background(
                                index % 2 == 0 ? Color.white : Color.gray.opacity(0.1)
                            )
                    }
                }
                .cornerRadius(8.0)
            }
```

Currently I found this bug where the scrollview stop working, I'm looking at it currently but not sure why. If you are available could you help me look. It's easier to replicate if you search multiple times.

https://user-images.githubusercontent.com/6356137/134111364-4731185b-c103-4041-9497-ee327aaa3871.mp4

Comment (vnntsu) [reply Source/Module/SearchRepository/SearchRepoScreen.swift:37]: @blyscuit I faced it too, but not much time to search and solve it tbh 🙈

Comment (vnntsu) [reply Source/Module/SearchRepository/SearchRepoScreen.swift:37]: > This block seems to work fine
> ```
> 
>             ScrollView(.vertical) {
>                 LazyVStack(alignment: .leading) {
>                     ForEach(Array(zip(viewModel.filteredItems.indices, viewModel.filteredItems)), id: \.0) { index, item in
>                         RepositoryCell(viewModel: item, isNotifying: notifiedReposObservable.getIsNotifying(item))
>                             .background(
>                                 index % 2 == 0 ? Color.white : Color.gray.opacity(0.1)
>                             )
>                     }
>                 }
>                 .cornerRadius(8.0)
>             }
> ```

really nice suggestion, it works perfectly, thank you <3

**Review by blyscuit — APPROVED** [review]

**Pull request merged**